readers/multishard.cc

```diff
@@ -864,9 +870,15 @@ future<> shard_reader::do_fill_buffer(std::optional<buffer_fill_hint> hint) {
                     return lifecycle_policy->create_reader(std::move(s), std::move(permit), pr, ps, std::move(ts), fwd_mr);
                 });
                 auto s = gs.get();
-                auto permit = co_await _lifecycle_policy->obtain_reader_permit(s, "shard-reader", timeout(), _trace_state);
+                auto permit_fut = co_await coroutine::as_future(_lifecycle_policy->obtain_reader_permit(s, "shard-reader", timeout(), _trace_state));
+                if (permit_fut.failed()) {
+                    co_return coroutine::return_exception_ptr(permit_fut.get_exception());
+                }
```

Comment:
Perphaps we can have

```c++
    auto permit = co_await coroutine::try_future(...);
```

which does what this sequence does: wait for the future, if it's successful unwrap it, if it failed then exit the coroutine.

It's telling the coroutine machinery: I did not manually wrap this future with try/catch, so you may as well complete the coroutine if it failed.

Also seems like the compiler should be able to do it - if it sees a throw (in future::get()), it can arrange for std::current_exception() to see it, and call coroutine_traits::unhandled_exception(). But that's too much to hope for.

Reply:
I will check out if this is doable. It would be much nicer to be sure.

Reply:
I managed to come up with something and it appears to be working well: https://github.com/scylladb/seastar/pull/2973
But there is a cost, I have to add an ugly callback to coroutine promise.
This is because calling `handle.destroy()` down-stack from `handle.resume()` results in double free of the coroutine frame. One can only destroy a suspended coroutine, so the caller of `handle.resume()` has to do it.